src/utilities/utilities.jl

```diff
@@ -242,15 +244,31 @@ function same_length_array(arr, value::Vector)
 end
 same_length_array(arr, value, key) = same_length_array(arr, convert_attribute(value, key))
 
+function to_ndim(T::Type{<: VecTypes}, vecs::AbstractArray{<: VecTypes}, fillval)
+    return map(vec -> to_ndim(T, vec, fillval), vecs)
+end
 function to_ndim(T::Type{<: VecTypes{N,ET}}, vec::VecTypes{N2}, fillval) where {N,ET,N2}
     T(ntuple(Val(N)) do i
         i > N2 && return ET(fillval)
         @inbounds return vec[i]
     end)
 end
 
+function to_ndim(trg::VT, src::VecTypes{N2}) where {N, N2, VT <: VecTypes{N}}
+    VT(ntuple(Val(N)) do i
+        @inbounds i > N2 ? trg[i] : src[i]
+    end)
+end
```

Comment:
I did some benchmarking on `to_ndim` while working on this pr. It's relatively slow if the type isn't known at compile time. You can see this by benchmarking the function directly:

```julia
T = Point4f
p2 = Point2f(1, 2)
fillval = 1f0
@benchmark Makie.to_ndim($T, $p2, $fillval) # ~74ns
```

And comparing with a wrapped version where the type is hardcoded:

```julia
f2(p, v) = Makie.to_ndim(Point4f, p, v)
@benchmark f2($p2, $fillval) # 1.4ns
```

Using a function barrier does not improve performance:

```julia
f1(T, p, v) = Makie.to_ndim(T, p, v)
@benchmark f1($T, $p2, $fillval) # ~74ns
```

However if you use a instantiated object rather than passing a type things become fast with a "dynamic" type:

```julia
p4 = Point4f(1)
@benchmark Makie.to_ndim($p4, $p2) # 1.4ns
```

This also has a real effect on `Makie.project`. With 1000 Point2f's it cuts down 150µs to about 9.6µs for me.